Email notifications for pipelines

ksierks opened the pull request:
## What does this PR do and why?
Implemented emails for notifying the submitter when a pipeline completes or errors.
Email template style copied from #464 for consistency. 
Addresses half of #476.

## Screenshots or screen recordings
![image](https://github.com/phac-nml/irida-next/assets/325703/5053529b-696c-4591-9879-b0b110ce7332)

![image](https://github.com/phac-nml/irida-next/assets/325703/fa8968f7-b5dc-4a7c-96bc-ca81c7acc814)

## How to set up and validate locally
1. Install local mock email service, [Mailhog](https://github.com/mailhog/MailHog).
1. After setup, open a terminal and run `~/go/bin/MailHog`.
1. Run irida-next.
1. Open a rails console.
1. Select a workflow execution.
`w = WorkflowExecution.first`
1. Make sure email notifications are enabled.
`w.email_notification = true`
1. Change state to `completed` or `error`.
`w.state = "completed"`
1. Save the workflow execution.
`w.save`
1. Check mock email service for sent email. Open a browser and navigate to http://0.0.0.0:8025/.
1. Repeat steps for state `error`.


Note: Email previews are located at http://localhost:3000/rails/mailers/pipeline_mailer.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

ChrisHuynh333 commented on app/views/mailers/pipeline_mailer/complete_email.html.erb:
```diff
@@ -0,0 +1,93 @@
+                  <div role="separator" style="line-height: 24px">&zwj;</div>
+                  <p class="m-0" style="line-height: 24px">
+                    <%= t(:".copy_paste") %>
+                    <a href="<%= workflow_execution_url(@workflow_execution) %>">
+                      <%= workflow_execution_url(@workflow_execution) %>
```
I think having this underlined since it's a link would be good.

ksierks replied:
2f949044248a246e6e9240fc5d17373d9f1f0881

ChrisHuynh333 reviewed (changes requested):
Hey Katherine, this is working as described. I looked at the emails in actual email clients and unfortunately some of the styling is lost:
Work outlook:
![image](https://github.com/phac-nml/irida-next/assets/82407232/464235eb-aea7-48b6-bec6-e85702153bd7)

Browser outlook:
![image](https://github.com/phac-nml/irida-next/assets/82407232/b341a43a-01c1-443d-a91f-082b01996858)

I talked to Deep about this and if it's possible to fix this with roadie, that'd be great. If you find it's not feasible, then an alternative may be needed.

ksierks commented:
> Hey Katherine, this is working as described. I looked at the emails in actual email clients and unfortunately some of the styling is lost: Work outlook: ![image](https://private-user-images.githubusercontent.com/82407232/320963282-464235eb-aea7-48b6-bec6-e85702153bd7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI4ODk1NzcsIm5iZiI6MTcxMjg4OTI3NywicGF0aCI6Ii84MjQwNzIzMi8zMjA5NjMyODItNDY0MjM1ZWItYWVhNy00OGI2LWJlYzYtZTg1NzAyMTUzYmQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEyVDAyMzQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVlOGQ5Mjc3NzA1ODFjMTUyYmYzYWM0NTI3MzViZTIxMzYyYmRjOGQ3MDZlMWVhYzk2ZWUyYjIyZWNlYmI0NGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._PvDtW9WcC5x35I_Z1CMihAig_-l6ULq-YevUyDavgk)
> 
> Browser outlook: ![image](https://private-user-images.githubusercontent.com/82407232/320963413-b341a43a-01c1-443d-a91f-082b01996858.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI4ODk1NzcsIm5iZiI6MTcxMjg4OTI3NywicGF0aCI6Ii84MjQwNzIzMi8zMjA5NjM0MTMtYjM0MWE0M2EtMDFjMS00NDNkLWE5MWYtMDgyYjAxOTk2ODU4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEyVDAyMzQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBhYzFjNzljYmU3YzkyN2NiNDQ0NzUxOTg3Yzg1OWFiMjMzYmUxMzJmMTFmNmM1MGE2MDg1YTU4MmJiZWM0OTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cgiudTDaqpCybf7JcsKXrOaI5_ubbIDMkH_S0GNRI_g)
> 
> I talked to Deep about this and if it's possible to fix this with roadie, that'd be great. If you find it's not feasible, then an alternative may be needed.

I think the problem might be the custom css in application.css. Importing a separate mailer.tailwind.css file might work.
For simplicity, I removed roadie-rails and stuck with the inline plain css and tried to use lots of tables as that's what Outlook likes. I still need to see what it looks like in other mail clients.

ericenns commented:
> I think the problem might be the custom css in application.css. Importing a separate mailer.tailwind.css file might work.
> For simplicity, I removed roadie-rails and stuck with the inline plain css and tried to use lots of tables as that's what Outlook likes. I still need to see what it looks like in other mail clients.

The problem was that outlook has very minimal CSS support, especially ignoring any styling inside of `<div>` tags. https://www.emailonacid.com/blog/article/email-development/how-to-code-emails-for-outlook-2016/

Though I did see some error logging from roadie-rails due to tailwindcss

ksierks commented:
> > I think the problem might be the custom css in application.css. Importing a separate mailer.tailwind.css file might work.
> > For simplicity, I removed roadie-rails and stuck with the inline plain css and tried to use lots of tables as that's what Outlook likes. I still need to see what it looks like in other mail clients.
> 
> The problem was that outlook has very minimal CSS support, especially ignoring any styling inside of `<div>` tags. https://www.emailonacid.com/blog/article/email-development/how-to-code-emails-for-outlook-2016/
> 
> Though I did see some error logging from roadie-rails due to tailwindcss

What I meant by the problem, was I was trying to fix the differences b/w the export outlook client and pipeline outlook client. There was some extra table bordering. I was only using Roadie to help emails render tailwindcss, nothing more. Assuming the error is the one in the CI logs, I forgot to commit the gem.lock file. I've removed Roadie completely.
From my PR description, the intention was to copy the style from the export email notification PR as it was ready for testing first. I spent a significant amount of time on my commit from yesterday trying to get the email to look good within the outlook email client as that is our main workplace email client. This is a screenshot with how it looks from an Outlook client with a zoom out of 60% to be able to see everything:
![email](https://github.com/phac-nml/irida-next/assets/325703/8d4d3852-b11a-41dd-8ac8-61257ab345be)
If this doesn't address the concern, I'll match what is done from the export email notification PR, thank you.

ChrisHuynh333 reviewed (approved):
Looks good to me. Email format looks great in the work email client!

deepsidhu85 commented on app/views/mailers/pipeline_mailer/complete_email.html.erb:
```diff
@@ -0,0 +1,98 @@
+            <tr>
+              <td style="padding:12px 24px;">
+                <p style="color:#334155;line-height:24px;">
+                  <%= t(:".body.start") %>
```
Can I get you to just have one translations to replace `.body.start` and `.body.end.` You can just pass in params to the translation. This is mainly due to translations sometimes get translated correctly when they are separated like this

ksierks replied:
I tried using one translation, but it looks like this if I do....
![image](https://github.com/phac-nml/irida-next/assets/325703/f52e48de-7b36-42e0-8932-01c93f38129a)

ChrisHuynh333 replied:
I fixed it in my emails PR. Add `_html` to your translation tag and it should convert the styling. Such as `ready_for_download_html: "Your export <b>%{name}</b> is ready for download."`

ksierks replied:
Right, thanks! ca992a0e2e94a9c4212b50bbf93f00b1be7b8ef1

deepsidhu85 commented on app/views/mailers/pipeline_mailer/error_email.html.erb:
```diff
@@ -0,0 +1,98 @@
+            <tr>
+              <td style="padding:12px 24px;">
+                <p style="color:#334155;line-height:24px;">
+                  <%= t(:".body.start") %>
```
Same thing applies as the above comment

ksierks replied:
ca992a0e2e94a9c4212b50bbf93f00b1be7b8ef1

deepsidhu85 reviewed (changes requested):
This is looking great. Just have 2 comments below about the translations for the completed and errored emails

deepsidhu85 reviewed (approved):
Looks good to me!
